man/en/jubaconv.1
.\" Man page generated from reStructuredText.
.
.TH "JUBACONV" "1" " " "" "Jubatus"
.SH NAME
jubaconv \- jubaconv
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
jubaconv [options ...]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
\fBjubaconv\fP is a tool to test your fv_converter configuration.
.sp
\fBjubaconv\fP simulates the internal behavior of fv_converter and displays the result of conversion on the command\-line.
.sp
As \fBjubaconv\fP only processes 1 record, \fBglobal_weight\fP of string feature extraction rules (\fBstring_rules\fP) will not be calculated.
Use \&../api/api_weight if you need a simulation including \fBglobal_weight\fP\&.
.SS Options
.INDENT 0.0
.IP \(bu 2
\fB[]\fP indicates the default value.
.UNINDENT
.INDENT 0.0
.TP
.B \-i <format>, \-\-input\-format <format>
Format of the input. [json]
.sp
\fB<format>\fP must be one of \fBdatum\fP or \fBjson\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-o <format>, \-\-output\-format <format>
Format of the output. [fv]
.sp
\fB<format>\fP must be one of \fBfv\fP, \fBdatum\fP or \fBjson\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-c <config>, \-\-conf <config>
Jubatus server configuration file in JSON.
.sp
This option must be given only if \fBfv\fP is specified for \fI\%\-o\fP\&.
.UNINDENT
.INDENT 0.0
.IP \(bu 2
.INDENT 2.0
.TP
.B Note: If the input format is \fBjson\fP, we have to add a slash symbol \fB/\fP at the start of \fBkey\fP in jubatus server config files.
.INDENT 7.0
.IP \(bu 2
If the input format is \fBdatum\fP: \fB"num_rules" : [{"key": "number", "type": "num"}]\fP
.IP \(bu 2
If the input format is \fBjson\fP:  \fB"num_rules" : [{"key": "/number", "type": "num"}]\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH FILE FORMATS
.SS Input
.sp
Supported formats are \fBdatum\fP and \fBjson\fP\&.
See the Examples section for details.
.INDENT 0.0
.IP \(bu 2
\fBdatum\fP is a Datum\-styled JSON data structure.
The root must have 3 keys (\fBstring_values\fP, \fBnum_values\fP and \fBbinary_values\fP), and whose values must be a flat object, whose keys are string and values are string (for \fBstring_values\fP and \fBbinary_values\fP) or integer/float (for \fBnum_values\fP).
.IP \(bu 2
\fBjson\fP is an arbitrary JSON data structure.
JSON data structure will be flatten to key\-value format by joining keys with \fB/\fP separator.
.UNINDENT
.SS Output
.sp
Supported formats are \fBfv\fP, \fBdatum\fP and \fBjson\fP\&.
.INDENT 0.0
.IP \(bu 2
\fBfv\fP prints the feature vectors extracted from the input record by fv_converter.
.IP \(bu 2
\fBdatum\fP prints the input record converted to Datum as Datum\-styled JSON data structure.
.IP \(bu 2
\fBjson\fP pretty\-prints the input record without any conversion.
This can only be used if the input format is \fBjson\fP\&.
.UNINDENT
.SH EXAMPLES
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ cat data.json
{ "message": "hello world", "age": 31 }

$ jubaconv \-i json \-o fv \-c /opt/jubatus/share/jubatus/example/config/classifier/pa.json < data.json
/message$hello world@str#bin/bin: 1
/age@num: 31

$ cat datum.json
{
  "string_values": {
    "hello": "world"
  },
  "num_values": {
    "age": 31
  },
  "binary_values": {
  }
}

$ jubaconv \-i datum \-o fv \-c /opt/jubatus/share/jubatus/example/config/classifier/pa.json < datum.json
hello$world@str#bin/bin: 1
age@num: 31
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
PFN & NTT
.SH COPYRIGHT
2011-2017 PFN & NTT
.\" Generated by docutils manpage writer.
.


ChangeLog.rst
Release 1.0.2 - 2017/02/27
--------------------------

* Changes to Supported Systems
    * Ubuntu 16.04 LTS is now officially supported (#1180, #1181)

Release 1.0.1 - 2016/12/26
--------------------------

* Improvements
    * Support OpenCV 2.0 in image extraction plugin (#1168, #1174, #1175)
    * fv_converter plugin loading code moved to Jubatus Core (#1171)
    * Improve Dockerfile (#1170, #1172, #1173)

Release 1.0.0 - 2016/10/31
--------------------------

* New Features
    * Add image extraction plug-in (#843, #1150)
    * Add configuration validation feature (#973, #1144)

* Improvements
    * clustering: change APIs to accept ID-based datum (#1155)
    * Update Dockerfile to use Ubuntu 14.04 (#1146)
    * Improve ``--daemon`` option of server and proxy to detach from console (#1072, #1148)
    * classifier: models are now compatible in standalone and distributed mode (#469, #1153)
    * Change the default log level to INFO instead of DEBUG (#1152)
    * Update example configurations (#1165)
    * Update manpages (#1167)
    * Update packaging tools (#1157)

* Bug fixes
    * graph: fix IDs being reused when loading model file to other server instance in standalone mode (#1149, #1156)

Release 0.9.4 - 2016/08/29
--------------------------

* Improvements
    * Improve error detection of log4cxx config file (#891, #1142)
    * Print DEBUG logs by default (#947, #1143)
    * Support printing logs in UTF-8 (#1078, #1140)
    * Improve comment in wscript (#1138, #1139)

* Bug fixes
    * Fix missing locks in mixer (#1092, #1135)
    * Fix RPM packaging tool to work on RHEL 6.x (#1137)

Release 0.9.3 - 2016/07/25
--------------------------

* Improvements
    * Improve behavior when ZooKeeper log files cannot be reopened (#1026, #1133)
    * Remove Glibc C++ debug flag in debug mode (#910, #1136)
    * Improve packaging tools to remove dependency to Google Code (#980, #1130)

* Bug fixes
    * weight: fix to display config on startup (#1131, #1132)
    * Fix missing lock in get_status of linear_mixer (#1093, #1134)

Release 0.9.2 - 2016/06/27
--------------------------

* Improvements
    * Update README.rst (#1126, #1129)

* Bug fixes
    * Fix `weight_client.hpp` not included in `client.hpp` (#1127, #1128)

Release 0.9.1 - 2016/05/30
--------------------------

* New feature
    * weight: server that returns raw results of fv_converter (#1105, #1116)

* Improvements
    * Add `default.json` config files (#1117, #1120)
    * Add example config file for Anomaly (lof with unlearner) (#1118)
    * Update manpage (#1123)

* Bug fixes
    * Fix thread safetiness issue in save/load RPC (#1113, #1115)
    * Fix `dl` link error on some distributions (#1121, #1122)

Release 0.9.0 - 2016/04/25
--------------------------

* New feature
    * `jubakit <https://github.com/jubatus/jubakit>`_: Python module to easily evaluate precision of models (#1044)

* Changes to Supported Systems
    * Ubuntu 14.04 LTS is now officially supported (#1071)

* Improvements
    * classifier: improve get_labels API to return number of trained instances (#1109)
    * Add example configuration file for Nearest Neighbor (multi-threaded LSH) (#1111)
    * Add example configuration file for Classifier (cosine / euclidean) (#1108)
    * Add example configuration file for Recommender (euclid_lsh with unlearner) (#1108)
    * Configuration in model is used when loading model file on startup (#1081, #1107)
    * Improve compatibility with FreeBSD 9 (#1099)
    * Update manpages (#1112)
    * Code cleanup (#1102)

* Bug fixes
    * Fix jenerator test (#1103)

Release 0.8.9 - 2016/03/28
--------------------------

* Improvements
    * Support extracting specific part-of-speech in MeCab plugin (#1073, #1098)
    * Improved proxy to log timeout errors (#1048, #1095)

* Bug fixes
    * Fix "jubactl -c save" not working (#1094)

Release 0.8.8 - 2016/02/29
--------------------------

* Improvements
    * Support extracting base form in MeCab plugin (#1074, #1085)
    * Compare configuration string semantically on load (#1080, 1087)
    * Add example configuration file for Recommender (inverted_index_euclid with unlearner) (#1090)
    * Fix email address of DEB package (#1082, #1088)

* Bug fixes
    * Fix printing logs multiple times after receiving SIGHUP (#962, #1084, #1086)

Release 0.8.7 - 2016/01/25
--------------------------

* Improvements
    * Add example configuration file for Recommender (MinHash with unlearner) (#1077)
    * Add example configuration file for Recommender and Anomaly (inverted_index_euclid) (#1079)

Release 0.8.6 - 2015/12/25
--------------------------

* Improvements
    * Support morpheme n-gram in mecab plugin (#1054, #1070)

* Bug fixes
    * Fix typos in exception and log messages (#1068, #1069)

Release 0.8.5 - 2015/11/30
--------------------------

* Improvements
    * improve client tests to work on Python 3 (#1043, #1063)
    * improve build script (#1064)
    * recommender: fix typo of calc_similarity function name (#1065)
    * recommender: add example config for LSH with unlearner (#1066)
    * update man pages (#1067)

* Bug fixes
    * Fix RPM package errors (#1059, #1060, #1061, #1062)

Release 0.8.4 - 2015/10/26
--------------------------

* Improvements
    * nearest_neighbor: improve analysis methods performance by eliminating giant lock (#1051)

* Bug fixes
    * Fix loading model files via command line option not working (#1055, #1057)

Release 0.8.3 - 2015/09/28
--------------------------

* Improvements
    * clustering: add seed option to example configuration files (#1047)
    * Fix packaging tools to use new oniguruma URL (#1045, #1046)

* Bug fixes
    * nearest_neighbor: fix argument type (#1042, #1049, #1050)

Release 0.8.2 - 2015/08/31
--------------------------

* Improvements
    * Add Dockerfile (#904)
    * Support add aggregator in jenerator (#1022, #1040)
    * Migrate ux-trie packaging from Google Code to GitHub (#1037)

* Bug fixes

jubatus/server/cmd/jubaconv.cpp
// Jubatus: Online machine learning framework for distributed environment
// Copyright (C) 2011 Preferred Networks and Nippon Telegraph and Telephone Corporation.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License version 2.1 as published by the Free Software Foundation.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

#include <log4cxx/basicconfigurator.h>
#include <log4cxx/patternlayout.h>
#include <log4cxx/fileappender.h>

#include <string>

#include "jubatus/util/text/json.h"

#include "jubatus/core/common/exception.hpp"
#include "jubatus/core/fv_converter/datum_to_fv_converter.hpp"
#include "jubatus/core/fv_converter/datum.hpp"
#include "jubatus/core/fv_converter/json_converter.hpp"
#include "jubatus/core/fv_converter/converter_config.hpp"
#include "jubatus/core/fv_converter/so_factory.hpp"
#include "../third_party/cmdline/cmdline.h"


using std::bad_cast;
using std::cerr;
using std::cin;
using std::cout;
using std::endl;
using std::ifstream;
using std::string;
using jubatus::util::text::json::json;
using jubatus::core::fv_converter::converter_config;
using jubatus::core::fv_converter::datum;
using jubatus::core::fv_converter::datum_to_fv_converter;
using jubatus::core::fv_converter::json_converter;

struct server_config {
  jubatus::core::fv_converter::converter_config converter;

  template<typename Ar>
  void serialize(Ar& ar) {
    ar & JUBA_MEMBER(converter);
  }
};

void show_invalid_type_error(
    const string& input_format,
    const string& output_format) {
  cerr << "invalid input-output type: " << input_format << " -> "
      << output_format << endl;
}

void output_json(const json& json) {
  json.pretty(cout, false);
}

void output_datum(const datum& datum) {
  jubatus::util::text::json::to_json(datum).pretty(cout, false);
}

int read_config(const string& conf_file, converter_config& conf) {
  ifstream ifs(conf_file.c_str());
  if (!ifs) {
    cerr << "cannot open converter config file: " << conf_file << endl;
    return -1;
  }
  server_config config;
  ifs >> jubatus::util::text::json::via_json(config);
  conf = config.converter;
  return 0;
}

void convert_datum(
    const datum& datum,
    jubatus::core::common::sfv_t& fv,
    const string& conf_file) {
  if (conf_file == "") {
    cerr << "specify converter config with -c flag" << endl;
    exit(-1);
  }
  datum_to_fv_converter conv;
  converter_config conf;
  jubatus::core::fv_converter::so_factory so_loader;
  read_config(conf_file, conf);
  initialize_converter(conf, conv, &so_loader);
  conv.convert(datum, fv);
}

void output_fv(const jubatus::core::common::sfv_t& fv) {
  for (size_t i = 0; i < fv.size(); ++i) {
    cout << fv[i].first << ": " << fv[i].second << endl;
  }
}

void read_json(jubatus::util::text::json::json& json) {
  try {
    cin >> json;
  } catch (const bad_cast& e) {
    cerr << "invalid json format" << endl;
    exit(-1);
  }
}

void read_datum(datum& datum) {
  try {
    cin >> jubatus::util::text::json::via_json(datum);
  } catch (const bad_cast& e) {
    cerr << "invalid datum format" << endl;
    exit(-1);
  }
}

void disable_log4cxx() {
    log4cxx::LayoutPtr layout(
        new log4cxx::PatternLayout(""));
    log4cxx::AppenderPtr appender(
        new log4cxx::FileAppender(layout, "/dev/null"));
    log4cxx::BasicConfigurator::configure(appender);
}

int main(int argc, char* argv[])
try {
  cmdline::parser p;
  p.add<string>("input-format", 'i', "input format (json/datum)",
      false, "json", cmdline::oneof<string>("json", "datum"));
  p.add<string>("output-format", 'o', "output format (json/datum/fv)", false,
      "fv", cmdline::oneof<string>("json", "datum", "fv"));
  p.add<string>("conf", 'c', "converter config file", false);
  p.set_program_name("jubaconv");
  p.parse_check(argc, argv);

  jubatus::util::text::json::json json;
  datum datum;
  jubatus::core::common::sfv_t fv;

  bool proc = false;
  string input_format = p.get<string>("input-format");
  string output_format = p.get<string>("output-format");

  disable_log4cxx();

  if (input_format == "json") {
    read_json(json);
    proc = true;
  }

  if (output_format == "json") {
    if (!proc) {
      show_invalid_type_error(input_format, output_format);
      return -1;
    }
    output_json(json);
    return 0;
  }

  if (input_format == "datum") {
    read_datum(datum);
    proc = true;
  } else if (proc) {
    json_converter::convert(json, datum);
  }

  if (output_format == "datum") {
    if (!proc) {
      show_invalid_type_error(input_format, output_format);
      return -1;
    }
    output_datum(datum);
    return 0;
  }

  if (proc) {
    convert_datum(datum, fv, p.get<string>("conf"));
  }

  if (output_format == "fv") {
    if (!proc) {
      show_invalid_type_error(input_format, output_format);
      return -1;
    }
    output_fv(fv);
    return 0;
  }

  return -1;
} catch (const jubatus::core::common::exception::jubatus_exception& e) {
  std::cout << e.diagnostic_information(true) << std::endl;
}
